persist collector data and implement the following:

robskrob opened the pull request:
- setup up gcloud for persisting document storage
- draw route and enpoint for POST /api/v1/collectors
- create validation for collector requests
- set up model layer that persists collector data
- provide validations for data to be persisted in model layer
- set up plumbing for writing requests specs

Commit:
persist collector data:
- setup up gcloud for persisting document storage
- draw route and enpoint for POST /api/v1/collectors
- create validation for collector requests
- set up model layer that persists collector data
- provide validations for data to be persisted in model layer
- set up plumbing for writing requests specs

robskrob commented:
@chadwpry this PR is ready for you when you can review it.

Commit:
update README to include instructions for POST a collector

chadwpry commented on README.md:
```diff
@@ -51,6 +51,23 @@ environment.
 
+### Collector Resource
+
+When send a POST request to create a collector, the request must have the below URI and include the following key/value pairs in the payload -- hostname, payload and payload['nonce']:
```
When sending?

chadwpry commented on README.md:
```diff
@@ -51,6 +51,23 @@ environment.
+
+    body: {
+     hostname=www.ikea.com,
+     payload={ nonce: 'kjhs9182791hakajh19',
```
Let's put the nonce at as a top level parameter instead of inside the payload. That way the payload can be created by the capture routine in the extension and passed through to the api without any changes.

chadwpry commented on models/index.js:
```diff
@@ -0,0 +1,59 @@
+    }, function (err) {
+      if (err) {
+        return errorCallback(err);
+      }
```
Use the else to handle success instead of being a fall through to line 55

chadwpry commented on models/index.js:
```diff
@@ -0,0 +1,59 @@
+  }
+}
+
+var model = {
```
Let's use individual function exports... e.g.

let savePayload = (payload, successCallback, errorCallback) => {
};

module.exports.savePayload = savePayload;

chadwpry commented on routes/api/collectors.js:
```diff
@@ -0,0 +1,74 @@
+var express = require('express');
```
Let's start to change vars to lets

In vim it looks like this...

:%s/var/let/

chadwpry commented on routes/api/collectors.js:
```diff
@@ -0,0 +1,74 @@
+var router = express.Router();
+var model = require('../../models/index');
+
+function validateProduct(payload) {
```
Let's remove validateProduct from the routing

chadwpry commented on routes/api/collectors.js:
```diff
@@ -0,0 +1,74 @@
+  }
+}
+
+function saveCollector(fields) {
```
Same here, include this in the persistence (model) layer instead of routes. It is not a concern of the routing, only persistence.

chadwpry commented on routes/api/collectors.js:
```diff
@@ -0,0 +1,74 @@
+}
+
+router.post('/', function(req, res, next) {
+  let promise = new Promise(function(resolve, reject){
```
Let's look at my PR as a model for this... I don't think the promise is necessary here, although we want that type of logic and will get it from the success and error callbacks in the function.

Commit:
remove validation logic from the routes layer, refactor validations
in model layer, refactor model layer to export gcloud datastore client and
savePayload function and respond to client with collectorKey id and kind.

Commit:
include config/* in .gitignore

robskrob commented on config/config.json:
```diff
@@ -0,0 +1,14 @@
+{
```
this is in .gitignore...not too sure why it's visible in the PR

robskrob commented on models/index.js:
```diff
@@ -0,0 +1,60 @@
+    if (err) {
+      errorCallback(err);
+    } else {
+      successCallback(collectorKey);
```
should I pass the recently created collector document to the successCallback, or does the collectorKey suffice? It contains the id and the kind attributes.

robskrob commented on routes/api/collectors.js:
```diff
@@ -0,0 +1,24 @@
+
+router.post('/', function(req, res, next) {
+  models.savePayload(req.body, (collectorKey) => {
+        res.json({
```
I wanted to make a serializer but noticed that I do not have a model and that the only data available to respond to the client with is the id and the kind attributes.

Commit:
create saveSample, remove index.js, set up seed data for
sample, refactor sample create endpoint and refactor
saveSupplier

chadwpry commented on README.md:
```diff
@@ -87,6 +87,24 @@ environment.
+
+*Request Example*
+
+    `POST /api/v1/collectors
```
This is a samples api now

Commit:
feedback for PR

chadwpry commented on package.json:
```diff
@@ -13,6 +13,7 @@
     "body-parser": "^1.15.2",
     "debug": "^2.2.0",
     "express": "^4.13.4",
+    "google-cloud": "^0.40.0",
```
Remove this from the package list

Commit:
google cloud is not used

chadwpry commented:
👍

chadwpry commented:
Issue #15 is covered by this PR
